lib/spack/spack/solver/asp.py

```diff
@@ -1546,8 +1556,13 @@ def key_fn(item):
                 # if there is a concrete version on the CLI *that we know nothing
                 # about*, add it to the known versions. Use idx=0, which is the
                 # best possible, so they're guaranteed to be used preferentially.
+                if i < len(specs):
+                    origin = version_provenance.spec
+                else:
+                    origin = version_provenance.dev_spec
+
```

Comment:
This is hard to read and it's hard to understand that you're basically doing the same thing to two lists with one tweak.

This should be refactored by pulling the loop nest out as its own function like `_add_concrete_versions(self, specs, origin)`, then refactor this block as:

```python
self._add_concrete_versions(specs, version_provenance.spec)
self._add_concrete_versions(dev_specs, version_provenance.dev_spec)
```